Fix Sitemap.asp

ExtremeFiretop opened the pull request:
In short; On 3004 firmware, the router’s JavaScript populates a global array myMenu as soon as you call show_menu().
Therefore, the original code just checked if(myMenu.length == 0) ... retry and then called GenerateSiteMap(...) once myMenu was non‐empty.

However, on 3006 firmware, ASUS changed how they store menu data. They use seem to now use:
`Session.set("menuList.", menuTree.list);`
in state.js rather than assigning to the global myMenu.

As a result, myMenu never gets populated automatically, so the old check (if(myMenu.length == 0) ... retry) would loop forever on 3006.

```
DEBUG: show_menu() is defined. Calling it now..
DEBUG: Now calling LoadSiteMap()...
DEBUG: Entering LoadSiteMap()...
DEBUG: menuList or menuExclude not ready; retry in 1s...
DEBUG: Added click event listener to #sitemap_showurls
DEBUG: Entering LoadSiteMap()...
DEBUG: menuList or menuExclude not ready; retry in 1s...
DEBUG: Entering LoadSiteMap()...
DEBUG: menuList or menuExclude not ready; retry in 1s..
DEBUG: Entering LoadSiteMap()..
DEBUG: menuList or menuExclude not ready; retry in 1s...
```

My PR works on both 3004 and 3006 by using a hybrid of both approaches:
First checks if myMenu is already populated (the 3004 way).
If not, it tries to pull menuList. from Session (the 3006 way).
Then it sets myMenu (and menuList, if needed) so GenerateSiteMap(...) will see the data regardless of firmware version.

Martinski4GitHub commented:
> In short; On 3004 firmware, the router’s JavaScript populates a global array myMenu as soon as you call show_menu(). Therefore, the original code just checked if(myMenu.length == 0) ... retry and then called GenerateSiteMap(...) once myMenu was non‐empty.
> 
> However, on 3006 firmware, ASUS changed how they store menu data. They use seem to now use: `Session.set("menuList.", menuTree.list);` in state.js rather than assigning to the global myMenu.
> 
> As a result, myMenu never gets populated automatically, so the old check (if(myMenu.length == 0) ... retry) would loop forever on 3006.
> 
> ```
> DEBUG: show_menu() is defined. Calling it now..
> DEBUG: Now calling LoadSiteMap()...
> DEBUG: Entering LoadSiteMap()...
> DEBUG: menuList or menuExclude not ready; retry in 1s...
> DEBUG: Added click event listener to #sitemap_showurls
> DEBUG: Entering LoadSiteMap()...
> DEBUG: menuList or menuExclude not ready; retry in 1s...
> DEBUG: Entering LoadSiteMap()...
> DEBUG: menuList or menuExclude not ready; retry in 1s..
> DEBUG: Entering LoadSiteMap()..
> DEBUG: menuList or menuExclude not ready; retry in 1s...
> ```
> 
> My PR works on both 3004 and 3006 by using a hybrid of both approaches: First checks if myMenu is already populated (the 3004 way). If not, it tries to pull menuList. from Session (the 3006 way). Then it sets myMenu (and menuList, if needed) so GenerateSiteMap(...) will see the data regardless of firmware version.
>
>
Great job, bud!! I'll take a closer look at your PR tomorrow evening.
I'm about to go to bed; I have a busy day tomorrow.
Have a good night!!

Martinski4GitHub commented:
@ExtremeFiretop,

Sorry, bud; I have not had the time to review & merge this PR.

Yesterday was Valentine's Day so my wife and I went for a romantic dinner and spent some extra time together, and today we had a "double date" with her sister & sister's husband, and we all went to have brunch at a French restaurant. The food was "très magnifique"!!! :>)

Anyway, I decided to spend whatever free time I have tonight on testing & validating the transition process from **1.3.10** to **1.4.0** versions of MerlinAU. IMO, this was more important than the scMerlin "Sitemap" issue.

I'll start to look at your PR tomorrow evening. Right now, my wife and I are going to watch a movie.

Have a good night, bud!!  Talk to you later.

ExtremeFiretop commented:
> @ExtremeFiretop,
> 
> Sorry, bud; I have not had the time to review & merge this PR.
> 

No worries, same with me to yours hahaha 😂 

> Yesterday was Valentine's Day so my wife and I went for a romantic dinner and spent some extra time together, and today we had a "double date" with her sister & sister's husband, and we all went to have brunch at a French restaurant. The food was "très magnifique"!!! :>)
> 

Ah love it! If it's traditional French food it would be fairly small portions but very good. Très délicieux!

> Anyway, I decided to spend whatever free time I have tonight on testing & validating the transition process from **1.3.10** to **1.4.0** versions of MerlinAU. IMO, this was more important than the scMerlin "Sitemap" issue.
> 

Fair point! I'll check into your PR in the morning my eyes are closing after a busy weekend myself. We've had snow storm after snow storm and over 27 inch's in snow fall over the last week. 

I recently purchased my girlfriend's first car; and it's a used car, but it's nice considering it's going to be her first car.

It's a 2014 Dodge Dart SXT 2.4L. Came with a safety, winter tires, delivered to the house on a flatbed with a full tank of gas. 

Only a handful of things to fix but that's what I've been busy with this weekend. Washer fluid pump didn't work so I replaced that so she can see while driving lol. (Surprisingly not apart of a safety check)

Did an oil change, replaced the cabin air filter. Gave it a clean wash inside and out. Some lights were acting funky, they worked when the daytime runner lights were off. But they would stop working when they were on, turns out it was a wiring issue I found in the truck of the car. 

Ran some new wire and soldered that in to fix the lighting issues, ran all that in some conduit to make it look nice and stock. 

> I'll start to look at your PR tomorrow evening. Right now, my wife and I are going to watch a movie.
> 
> Have a good night, bud!! Talk to you later.

Enjoy the movie and night bud! Don't forget my offer for movie streaming 🫴 Hehee

ExtremeFiretop commented:
@Martinski4GitHub 

Friendly check in :)

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> Friendly check in :)
>
Yeah, I plan to take a look and review it over the weekend. I have had to take care of other things first.

ExtremeFiretop commented:
@Martinski4GitHub 

On an unrelated note, I noticed that spdmerlin only currently supports the OpenVPN VPN interfaces?
How are would it be to implement the WireGuard VPN interfaces for speedtesting?

![image](https://github.com/user-attachments/assets/6603423d-365c-41fb-9987-95ed706b1459)

Don't worry about this PR specifically; your last message has brought me up to speed on how busy you've been!
Just figured I'd advise I tested this on my Gnuton router and my 3006 router without issue :)

ExtremeFiretop commented:
@Martinski4GitHub 

Is that a feature request I can make between buddies/pals? 
Or should I just fork it and start ripping it a new one myself? xD

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> On an unrelated note, I noticed that spdmerlin only currently supports the OpenVPN VPN interfaces?
> 
>
> How hard would it be to implement the WireGuard VPN interfaces for speedtesting?
> 
>
I don't really use spdMerlin at all, and the only changes I have made to the code were to fix compatibility issues with new F/W, or some glitches in the CLI or WebGUI. Other than that, I haven't reviewed that code very closely to have an assessment of the level of difficulty to add WireGuard. My guess would be a medium-level effort, but again, this is just a guess.

>
>
> 
> Don't worry about this PR specifically; your last message has brought me up to speed on how busy you've been! Just figured I'd advise I tested this on my Gnuton router and my 3006 router without issue :)
>
>
Yeah, honestly, this "Sitemap" issue is at the bottom of my current to-do list.  Looks like very few people really use it, and it doesn't affect the functionality of the core scMerlin add-on, so other things get higher priority in my limited spare time. 

But it's good to know that you have verified it on Gnuton as well.

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> Is that a feature request I can make between buddies/pals? Or should I just fork it and start ripping it a new one myself? xD
>
>
At this point, I'd recommend forking the code to at least take a look at the code to have a better idea of what's at stake. and what kind of effort it would take.

ExtremeFiretop commented:
> > @Martinski4GitHub
> > Is that a feature request I can make between buddies/pals? Or should I just fork it and start ripping it a new one myself? xD
> 
> At this point, I'd recommend forking the code to at least take a look at the code to have a better idea of what's at stake. and what kind of effort it would take.

I can do that!

I'm assuming it would be too too hard... Like you said probably a medium level effort. If it's not something your too familiar with odd the top of your head I can always take a stab at ripping open the code and checking it out.

I would love to see this for spdmerlin and I realize when it was originally wrote that OpenVPN was the standard (if not the only?) VPN client but at this point I think Wireguard gets lots of attention.

ExtremeFiretop commented:
If I happen to get something working, I'll submit a PR!

ExtremeFiretop commented:
So far this has been stupidly easy; I'm waiting for the hickup to show up:

![image](https://github.com/user-attachments/assets/e48ff477-7d5b-4026-9964-1bfa42578b99)

Martinski4GitHub commented:
> If I happen to get something working, I'll submit a PR!
>
>Yeah, submit a PR directly to JackYaz's repository; but keep in mind that you may not get a reply from him right away. In the past, usually, he would respond within a week, but for the last year, it has been months before he acknowledged the PR and responded.

ExtremeFiretop commented:
> > If I happen to get something working, I'll submit a PR!
> 
> Yeah, submit a PR directly to JackYaz's repository; but keep in mind that you may not get a reply from him right away. In the past, usually, he would respond within a week, but for the last year, it has been months before he acknowledged the PR and responded.

Will do!

![image](https://github.com/user-attachments/assets/7db1588c-fe52-40e1-8c78-68ef942f4db2)
![image](https://github.com/user-attachments/assets/ffd1d38b-ff6d-4c0e-8499-835b50a5d78f)

ExtremeFiretop commented:
Opened a PR directly to @decoderman

https://github.com/decoderman/scMerlin/pull/12

Martinski4GitHub commented:
> Opened a PR directly to @decoderman
> 
> [decoderman#12](https://github.com/decoderman/scMerlin/pull/12)
>
Good thinking! And yeah, that would be better since I likely won't have enough time to review it until other higher-priority items are off my to-do list. There are not enough hours in the day!! LOL!! :>).

ExtremeFiretop commented:
> > Opened a PR directly to @decoderman
> > [decoderman#12](https://github.com/decoderman/scMerlin/pull/12)
> 
> Good thinking! And yeah, that would be better since I likely won't have enough time to review it until other higher-priority items are off my to-do list. There are not enough hours in the day!! LOL!! :>).

I realized that he would be the repo published to AMTM, and while I literally have Zero issues with giving you the fix and credit. Maybe it's best I direct the fix to the right repo hehehe.

I literally just woke up now as I'm getting all the messages 

I did get @Jack to respond this morning!

ExtremeFiretop commented:
Just learned my email is being targeted by some kinda brute force attack.

![image001](https://github.com/user-attachments/assets/ce5b9aed-e7ac-43bd-929c-756188f35dc4)

Been going on for some time to what I can tell lol!

I'd love to see the person try, my password strength is calculated at taking 73 trillion years to brute force. 😂

Martinski4GitHub commented:
> Just learned my email is being targeted by some kinda brute force attack.
> 
> ![image001](https://private-user-images.githubusercontent.com/1971404/418321067-ce5b9aed-e7ac-43bd-929c-756188f35dc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4OTExNzgsIm5iZiI6MTc0MDg5MDg3OCwicGF0aCI6Ii8xOTcxNDA0LzQxODMyMTA2Ny1jZTViOWFlZC1lN2FjLTQzYmQtOTI5Yy03NTYxODhmMzVkYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDJUMDQ0NzU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTg2NDg2Y2FjY2FlZmFlZjUyYWI2YzVmZTRiNDhmMjkxNGZiYmU5NDk3ZjgzZTk4MWNiMGUyN2M3NjVkNjhmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.z-gBUE3IQsa1ehRvpyEnBQ-e9_x0hiO1rMNQ0qC3Tao)
> 
> Been going on for some time to what I can tell lol!
> 
>
The attacker(s) is/are bouncing off different countries (likely many different VPN servers) all over the world for the last 24 hours (at least). Good thing the Microsoft account notifies you of each event.

>
>
> I'd love to see the person try, my password strength is calculated at taking 73 trillion years to brute force. 😂
>
>
Yeah, all my passwords are at least 20 chars long, and 30 chars for all important sites (e.g. banks, credit card accounts, email accounts, etc.) plus 2-factor authentication, where available.

ExtremeFiretop commented:
> > Just learned my email is being targeted by some kinda brute force attack.
> > ![image001](https://private-user-images.githubusercontent.com/1971404/418321067-ce5b9aed-e7ac-43bd-929c-756188f35dc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4OTExNzgsIm5iZiI6MTc0MDg5MDg3OCwicGF0aCI6Ii8xOTcxNDA0LzQxODMyMTA2Ny1jZTViOWFlZC1lN2FjLTQzYmQtOTI5Yy03NTYxODhmMzVkYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDJUMDQ0NzU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTg2NDg2Y2FjY2FlZmFlZjUyYWI2YzVmZTRiNDhmMjkxNGZiYmU5NDk3ZjgzZTk4MWNiMGUyN2M3NjVkNjhmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.z-gBUE3IQsa1ehRvpyEnBQ-e9_x0hiO1rMNQ0qC3Tao)
> > Been going on for some time to what I can tell lol!
> 
> The attacker(s) is/are bouncing off different countries (likely many different VPN servers) all over the world for the last 24 hours (at least). Good thing the Microsoft account notifies you of each event.
> 
> > I'd love to see the person try, my password strength is calculated at taking 73 trillion years to brute force. 😂
> 
> Yeah, all my passwords are at least 20 chars long, and 30 chars for all important sites (e.g. banks, credit card accounts, email accounts, etc.) plus 2-factor authentication, where available.

That's exactly it. It's not a password it's a passphrase, it's long, it's complex, and same as you I have multifactor enabled on all the accounts when possible so I'm not worried but yay to being a target? 😁 HAHAHA

Funny someone out there thinks this is worth his time of a month+ twice every 2 hours. Obviously I'm in some kinda script but I'll be dead and he'll be dead and the PC will be dead long before it gets anywhere

Martinski4GitHub commented:
> > > Just learned my email is being targeted by some kinda brute force attack.
> > > ![image001](https://private-user-images.githubusercontent.com/1971404/418321067-ce5b9aed-e7ac-43bd-929c-756188f35dc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4OTExNzgsIm5iZiI6MTc0MDg5MDg3OCwicGF0aCI6Ii8xOTcxNDA0LzQxODMyMTA2Ny1jZTViOWFlZC1lN2FjLTQzYmQtOTI5Yy03NTYxODhmMzVkYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDJUMDQ0NzU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTg2NDg2Y2FjY2FlZmFlZjUyYWI2YzVmZTRiNDhmMjkxNGZiYmU5NDk3ZjgzZTk4MWNiMGUyN2M3NjVkNjhmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.z-gBUE3IQsa1ehRvpyEnBQ-e9_x0hiO1rMNQ0qC3Tao)
> > > Been going on for some time to what I can tell lol!
> > 
> > 
> > The attacker(s) is/are bouncing off different countries (likely many different VPN servers) all over the world for the last 24 hours (at least). Good thing the Microsoft account notifies you of each event.
> > > I'd love to see the person try, my password strength is calculated at taking 73 trillion years to brute force. 😂
> > 
> > 
> > Yeah, all my passwords are at least 20 chars long, and 30 chars for all important sites (e.g. banks, credit card accounts, email accounts, etc.) plus 2-factor authentication, where available.
> 
> That's exactly it. It's not a password it's a passphrase, it's long, it's complex, ...
>
Yes, password length and password entropy are the 2 key factors that determine strength.

>
> ... and same as you I have multifactor enabled on all the accounts when possible so I'm not worried but yay to being a target? 😁 HAHAHA
> 
> Funny someone out there thinks this is worth his time of a month+ twice every 2 hours. Obviously I'm in some kinda script but I'll be dead and he'll be dead and the PC will be dead long before it gets anywhere
>
>
LOL!!!
